### PR TITLE
libswift: reimplement `AssumeSingleThreaded` pass

### PR DESCRIPTION
**Pull request by MaxDesiatov**
I don't think that `AssumeSingleThreaded` is widely used, certainly doesn't look like a mandatory pass, and it seemed trivial enough for a first try of rewriting passes in Swift.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Review by eeckstein — APPROVED** [review]
nice!

File: lib/SILOptimizer/Transforms/AssumeSingleThreaded.cpp [hunk 1]
```diff
@@ -1,59 +0,0 @@
...
-// To get read of most atomic reference counting operations, the standard
-// library should be compiled in this mode as well 
-//
-// This pass affects only reference counting operations resulting from SIL
-// instructions. It wouldn't affect places in the runtime C++ code which
-// hard-code calls to retain/release. We could take advantage of the Instruments
-// instrumentation stubs to redirect calls from the runtime if it was
-// significant, or else just build a single-threaded variant of the runtime.
-//
-//===----------------------------------------------------------------------===//
```
Comment (benrimmington) [inline lib/SILOptimizer/Transforms/AssumeSingleThreaded.cpp:26]: Should any of this documentation be moved into the new file?

**Comment by MaxDesiatov** [comment]
I've also modified `test/sil-passpipeline-dump/basic.test-sh` to exclude `"sil-assume-single-threaded"` from expected pipeline dump output. This pass was previously always enabled, but conditionally updated instructions based on the CLI flag. I think it's cleaner if the pass is disabled completely unless the CLI flag is specified.

File: lib/SILOptimizer/Transforms/AssumeSingleThreaded.cpp (hunk 1 above)
Comment (MaxDesiatov) [reply lib/SILOptimizer/Transforms/AssumeSingleThreaded.cpp:26]: Thanks, great catch! That's fixed now.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please test Windows platform

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AssumeSingleThreaded.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ // To get rid of most atomic reference counting operations, the standard
+ // library should be compiled in this mode as well .
+ //
+ // This pass affects only reference counting operations resulting from SIL
+ // instructions. It wouldn't affect places in the runtime C++ code which
+ // hard-code calls to retain/release. We could take advantage of the Instruments
+ // instrumentation stubs to redirect calls from the runtime if it was
+ // significant, or else just build a single-threaded variant of the runtime.
+ //
+ //===----------------------------------------------------------------------===//
```
[the remark this answers is not in the document]
Comment (MaxDesiatov) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AssumeSingleThreaded.swift:26]: There's `test/SILOptimizer/assume_single_threaded.sil`, which is passing with the new implementation.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Pull request merged**